webapp/app.py

```diff
@@ -143,4 +166,17 @@ def login_handler():
 @app.route("/demo")
 @login_required
 def demo():
+    authentication_token = flask.session["authentication_token"]
+    authorization_header = {
+        "Authorization": f"macaroon root={authentication_token}"
+    }
+    response = requests.get(
+        url=f"{ANBOXCLOUD_API_BASE}1.0/instances",
+        headers=authorization_header,
+    )
+    try:
+        response.raise_for_status()
+    except requests.exceptions.HTTPError as e:
+        error = e.response.json()
+        return flask.render_template("401.html", error=error["error"])
```

Comment:
You should make sure you're actually checking for a 401 response, and only handling it in that specific case. Any other errors should be re-raised. And you should use Flask's [default error page mechanism](https://flask.palletsprojects.com/en/1.1.x/errorhandling/#registering), not `render_template`.

``` python3
except requests.exceptions.HTTPError as http_error:
    if http_error.response.status_code == 401:
        flask.abort(401, http_error.response.json()["error"])
    else:
        raise http_error
```

Then you need to define elsewhere:

``` python3
@app.errorhandler(401)
def handle_bad_request(e):
    return flask.render_template("401.html", error=error["error"]), 401
```